controller/src/beerocks/master/son_master_thread.cpp

```diff
@@ -373,6 +373,105 @@ bool master_thread::handle_cmdu_1905_autoconfiguration_search(Socket *sd,
     return son_actions::send_cmdu_to_agent(sd, cmdu_tx);
 }
 
+/**
+ * @brief Encrypt the config data using AES and add to the WSC M2 TLV
+ *        The encrypted data length is the config data length padded to 16 bytes boundary.
+ *
+ * @param[in] m2 WSC M2 TLV
+ * @param[in] config_data config data in network byte order (swapped)
+ * @param[in] authkey 32 bytes calculated authentication key
+ * @param[in] keywrapkey 16 bytes calculated key wrap key
+ * @return true on success
+ * @return false on failure
+ */
+bool master_thread::autoconfig_wsc_add_m2_encrypted_settings(
+    std::shared_ptr<ieee1905_1::tlvWscM2> m2, WSC::cConfigData &config_data, uint8_t authkey[32],
+    uint8_t keywrapkey[16])
+{
+    // Calculate length of data to encrypt
+    // (= plaintext length + 32 bits HMAC aligned to 16 bytes boundary)
+    size_t len             = (config_data.getLen() + 8 + 16) & ~0xFU;
+    uint8_t encrypted[len] = {0};
+
+    auto encrypted_settings = m2->create_encrypted_settings();
+    if (!encrypted_settings)
+        return false;
+    if (!encrypted_settings->alloc_encrypted_settings(len))
+        return false;
+    if (!m2->add_encrypted_settings(encrypted_settings))
+        return false;
+
+    std::copy_n(config_data.getStartBuffPtr(), config_data.getLen(), encrypted);
+    uint8_t *kwa = &encrypted[config_data.getLen()];
+    // Add KWA which is the 1st 64 bits of HMAC of config_data using AuthKey
+    if (!mapf::encryption::kwa_compute(authkey, encrypted, config_data.getLen(), kwa))
+        return false;
+    std::copy_n(kwa, m2->authenticator().data_length, m2->authenticator().data);
```

Comment:
I only notice now...

The authenticator that you need to fill in here is the key_wrap_authenticator, not the global authenticator. The global authenticator follows the same algorithm to compute it, but over the entire M1 + M2 (excluding the authenticator itself, obviously) instead of only the encrypted settings. Cfr. page 48 of the WSC spec: "M2 = Version || N1 || N2 || Description || PKR [ || ConfigData ] || HMACAuthKey(M1 || M2*)" (M2* == M2 excluding the authenticator itself).

So, obviously this shouldn't be done here, but in a separate step after M2 is completely filled in.

Also note that on the slave side, this means you need to keep a copy of M1 around to be able to calculate the hmac (you can't keep a partially calculated hmac because you don't have the shared secret yet at the time M1 is sent, and the authkey is mixed in with the message to get an HMAC).

Reply:
Right - opened #158 for it, will fix in a separate PR.